src/System/Console/GetOpt/Generics.hs

```diff
@@ -118,17 +135,22 @@ mkEmptyArguments fields = case (sing :: Sing xs, fields) of
 -- * showing help?
 
 helpWrapper :: (All Option xs) =>
-  [String] -> NP FieldInfo xs -> a -> Either (IO ()) a
-helpWrapper args fields a =
+  String -> [Hint] -> [String] -> NP FieldInfo xs -> Either [String] a -> Result a
+helpWrapper header hints args fields result =
     case getOpt Permute [helpOption] args of
-      ([], _, _) -> Right a
-      (() : _, _, _) -> Left $ do
-        progName <- getProgName
-        let header = progName
-        putStrLn (usageInfo header (mkOptDescrs fields))
+      ([], _, _) -> case result of
```

Comment:
Your last point is a very good one. It's a bug. Fixed in the newest commit.

Also I hope the function is a bit easier to understand.